# Practice Notebook — Session 5 (Functions, Derivatives, Integrals)
Problems with answers for calculus and multivariable basics.

Generated: 2025-09-05T16:27:08.529306 UTC

---

## Problem 1 — Intervals
Write closed, open, and half-open intervals between a=2 and b=5.

In [1]:
print('Closed: [2,5]')
print('Open: (2,5)')
print('Half-open: [2,5) and (2,5]')

Closed: [2,5]
Open: (2,5)
Half-open: [2,5) and (2,5]


**Answer (computed):** Closed [2,5], Open (2,5), Half-open [2,5), (2,5]

## Problem 2 — Vector and Norms
Given v = [3,4,5], compute its L2 norm.

In [2]:
import numpy as np
v = np.array([3,4,5])
print('v:', v)
print('L2 norm:', np.linalg.norm(v))

v: [3 4 5]
L2 norm: 7.0710678118654755


**Answer (computed):** L2 norm = 7.0710678118654755

## Problem 3 — Derivatives
Compute derivative of f(x) = x^2 + 5x + 6 and second derivative of f(x) = sin(x) + x^3.

In [3]:
import sympy as sp
x = sp.symbols('x')
f1 = x**2 + 5*x + 6
f1p = sp.diff(f1, x)
print('f1\' =', f1p)
f2 = sp.sin(x) + x**3
f2pp = sp.diff(f2, x, 2)
print('f2 second derivative =', f2pp)

f1' = 2*x + 5
f2 second derivative = 6*x - sin(x)


**Answer (computed):** f1' = 2*x + 5, f2'' = 6*x - sin(x)

## Problem 4 — Limits
Compute lim_{x->0} sin(x)/x.

In [4]:
import sympy as sp
x = sp.symbols('x')
print('limit sin(x)/x as x->0 =', sp.limit(sp.sin(x)/x, x, 0))

limit sin(x)/x as x->0 = 1


**Answer (computed):** 1

## Problem 5 — Partial Derivatives
For f(x,y) = x^3 + 2x^2 y + y^2 + 2x + 1, compute ∂f/∂x and ∂f/∂y.

In [5]:
import sympy as sp
x,y = sp.symbols('x y')
f = x**3 + 2*x**2*y + y**2 + 2*x + 1
print('df/dx =', sp.diff(f, x))
print('df/dy =', sp.diff(f, y))

df/dx = 3*x**2 + 4*x*y + 2
df/dy = 2*x**2 + 2*y


**Answer (computed):** df/dx = 3*x**2 + 4*x*y + 2, df/dy = 2*x**2 + 2*y

## Problem 6 — Jacobian and Hessian
Compute Jacobian of F = [x^2 y, 5x + sin y] and Hessian of f(x,y) = -7x^2 -5xy -4y^2.

In [6]:
import sympy as sp
x,y = sp.symbols('x y')
F = sp.Matrix([x**2 * y, 5*x + sp.sin(y)])
J = F.jacobian([x,y])
print('Jacobian:\n', J)
f = -7*x**2 -5*x*y -4*y**2
H = sp.hessian(f, (x,y))
print('\nHessian:\n', H)

Jacobian:
 Matrix([[2*x*y, x**2], [5, cos(y)]])

Hessian:
 Matrix([[-14, -5], [-5, -8]])


**Answer (computed):** Jacobian =
```
Matrix([[2*x*y, x**2], [5, cos(y)]])
```
Hessian =
```
Matrix([[-14, -5], [-5, -8]])
```

## Problem 7 — Integrals
Compute indefinite integral of x^2 + 1 and definite integral from 0 to 2.

In [7]:
import sympy as sp
x = sp.symbols('x')
F = sp.integrate(x**2 + 1, x)
res = sp.integrate(x**2 + 1, (x,0,2))
print('Indefinite:', F)
print('Definite 0..2:', res)

Indefinite: x**3/3 + x
Definite 0..2: 14/3


**Answer (computed):** Indefinite integral = x**3/3 + x, Definite = 14/3